feat: decouple loading of planned items

weph opened the pull request:
Hey @ivan-lednev 

here's my proposal for #282

I tried the suggestion from your last comment (https://github.com/ivan-lednev/obsidian-day-planner/issues/282#issuecomment-1804535147). It worked, but things got slower as I added more daily notes with more list items. The same happened with an increasing number of tasks. You already mentioned that you use lists for everything, so I wanted to come up with something that works long-term and with large vaults.

To me, the obvious solution was to load only what's necessary. I came up with this abstraction `PlannedItems` that takes care of that. You can request items for a given day and get a readable store in return that will eventually contain (meaning once loaded) the items. You can injects loaders into `PlannedItems` to do the actual data loading. There's one for scheduled tasks, and one for daily notes list items and unscheduled tasks, as well as some helpers. The actual data loading is debounced so items for the week planner are loaded in one fell swoop instead of seven queries. Once data is loaded, it will be returned without requerying. If there's a metadata change, stores that have a subscription will be updated. Unsubscribed stores with a stale cache will be updated once there's a new subscription.

I messed up a little with the integration, though. I hadn't updated my fork since I started working on this and when I fetched your changes I realized that you had refactored quite a bit. The way it is currently integrated is closer to how the codebase was a couple of weeks ago. My goal for the moment was to show you a working version so you can decide what to do with it.

I'm open to suggestions and happy to help.

weph committed:
feat: decouple loading of planned items

weph committed:
fix: only use list items from the daily notes of configured heading

weph committed:
refactor: daily notes item loader for readability

weph committed:
refactor: scheduled tasks loader for readability

ivan-lednev commented:
Hello, @weph,

Thanks for the submission! Once I have the time, I'll look into it.

weph commented:
Hey @ivan-lednev 

I just rebased onto main and resolved the conflicts, keeping things the way I implemented them in my original draft. I saw that you pulled up dependencies from the task container and changed the data loading. So, you probably won't like it the way it is right now. I'll have to rewrite some code to make it fit in there nicely, though, as it's currently more tailored to the way the plugin was when I started working on this. Before I do that, please let me know if you're interested in this PR at all. I know I'm not the only one who liked to use list items, but it's your plugin and therefore your call.

ivan-lednev commented:
Hello, @weph,

First of all, thank you for taking the time to contribute to the project!

Your approach is to pull stuff from Dataview piece by piece, as needed, instead of getting it all together at once and then filtering it as needed. Is this a fair summary?

I see two issues with this solution:

1. This code is a whole additional layer of complexity. It's harder for me to reason about, and therefore maintain. Right now, all the data is a simple chain of Svelte stores. 

2. It adds specificity to the way we interact with Dataview. Here is how it may become problematic: right now I'm working on a time tracking feature inspired by org-mode and Toggl. It's going to have the following syntax:
```md
- [ ] PR review
      [clocked::2023-12-12T16:03:38--2023-12-12T16:21:08]
      [clocked::2023-12-12T18:10:38--2023-12-12T18:26:17]
```

It looks like with your approach I'll have to come up with some variation of this method:
```ts
  private taskQueryFor(days: Set<Day>): string {
    const conditions = Array.from(days.values())
      .map((d) => `scheduled = date(${d.asIso()})`)
      .join(" OR ");

    return `TASK WHERE ${conditions}`;
  }
```
instead of working with one regular JavaScript array.

Before I dive deeper into reviewing your PR, please share your thoughts on these alternatives:
1. Make list items optional, warn the users that it might hurt performance and point them to the docs explaining how to narrow down the Dataview search with a filter.
2. Since Dataview is also open-source, we can split the heavy `dataview.tasks` operation the way Obsidian splits its search. This way, even if it takes a lot of time, it's not going to block the UI.
4. The creator of Dataview [is working on a successor](https://github.com/blacksmithgu/datacore) that promises to have much faster queries, so sitting around and waiting is also an option 😄.

weph commented:
Hey @ivan-lednev 

yes, that is a pretty good summary.

The problem I had with the "load all and process later" approach was that as I added more tasks and list items to my development vault, Obsidian became laggy and unresponsive. After the initial load, changing a note would trigger a reload, making the UI freeze again.

Decoupling the loading from the UI or having a faster version of dataview would, of course, mitigate that problem. I understand that loading all the data is a convenient approach, but I do think loading only what's necessary is the better way. Having said that, this is just my view and it may not align with yours or what you have in mind for your plugin.

To give you clear answers on your three alternatives:

1. This would probably be the way to go if you don't want to deal with the complexity of this PR. If you want to go that route, you can close this PR and I'll try to come up with something that fits your approach.
2. It's an option, but not something I would pursue
3. No problem for me, since I'm currently running my fork of your plugin anyway. The people who are left behind might feel differently, but it's FOSS after all.

ivan-lednev commented:
@weph, I started looking into your PR more closely and experimenting with the solution, and I found the following issue:

The optimization you propose hangs on the assumption that this query (taken from [here](https://github.com/ivan-lednev/obsidian-day-planner/pull/328/files#diff-94ec1430c2fd4e2256a6510eab34926ac5f12ce43e1f6fecf6e2309a578ca45eR35)):
```ts
app.plugins.plugins.dataview.api.query(`TASK WHERE scheduled = date(2023-01-01)`)
```
is faster than the current implementation.

However I started this on my real vault (~3500 files, lots of tasks & list items):
```ts
setInterval(
  async () =>
    await app.plugins.plugins.dataview.api.query(
      `TASK WHERE scheduled = date(2023-01-01)`,
    ),
  1000,
);
```
and each iteration takes around half a second. So it's just as slow as pulling everything at once. I think such a query is internally the same as `dataview.pages(``).where(() => ...)`, i.e. the same I use right now.

I know only one way to make queries faster: narrow the Dataview source, since it caches folders and tags: https://blacksmithgu.github.io/obsidian-dataview/reference/sources/. Unfortunately, it doesn't allow filtering by property key/value.

Have you checked the latest plugin version on your development vault (0.16.0)? I did the same optimization of removing repeated Dataview queries, so it should be much faster just because now there is no 'query-dataview-for-each-open-column'.

So my guess is that you've noticed performance improvements just because you removed query repetition in your implementation, not because you used a faster Dataview API.

Please correct me if I'm missing something. If your solution is more performant, I'm ready to continue the review, otherwise, option 1 looks like the way to go.

weph commented:
Hey @ivan-lednev 

yes, the main idea was to limit scope (i.e. the amount of data/notes to query) and use caching to prevent unnecessary reloading. Apparently, I really have missed that loading all tasks is as fast as loading all scheduled tasks for one or a couple of days. 

So I looked into your current version, and indeed, it performed way better than previous versions, but I still noticed that there was something lagging. Since it wasn't the loading, I figured it had to be the filtering. I added profiling to `useVisibleTasks` and compared it to my version, and another version of mine with further optimizations. I took into account, that `useVisibleTasks` also maps data via `mapToTasksForDay`, so I profiled that in my version too.

To make it easier for you to compare the results, I uploaded my demo vault here: [demo-vault.tar.gz](https://github.com/ivan-lednev/obsidian-day-planner/files/13695071/demo-vault.tar.gz)

It contains 4,000 notes (in total 10,000 scheduled tasks, 10,000 unscheduled tasks, 10,000 day planner list items, 10,000 unrelated list items).

Here are the versions I compared and their loading/processing times (taken on a M1 Max) when switching from one week to the previous/next one in the weekly planner. Numbers were taken in a "warmed up" state, meaning Obsidian and Dataview had finished indexing, and the initial task loading was done (in case of your and my optimized version). I also took the time it takes after a metadata change.

## Your latest version with profiling added to `useVisibleTasks`
https://github.com/weph/obsidian-day-planner/tree/comparison/v0.16.0-with-profiling

- Switching week: ~210 ms
- After metadata change: ~310 ms (~100 ms for reloading tasks, ~210 for processing)


## My original PR:
https://github.com/weph/obsidian-day-planner/tree/comparison/weph-pr

- Switching week: ~200 ms (~190 ms for loading tasks, ~10 ms for loading daily note list items)
- After metadata change: ~200 ms (~190 ms for loading tasks, ~10 ms for loading daily note list items)


## My PR with optimizations:
https://github.com/weph/obsidian-day-planner/tree/comparison/weph-pr-optimized

- Switching week: 3.6 ms (~0.1 ms for "loading" tasks, ~1 ms for loading daily note list items, ~2.5 ms for mapping data)
- After metadata change: ~168 ms (~150 ms for reloading tasks, ~15 ms for reloading daily notes, ~3 ms for mapping data)

---

I also created a vault with twice the amount of data (8,000 notes, 20,000 scheduled tasks/unscheduled tasks/day planner list items/unrelated list items): [demo-vault-2.tar.gz](https://github.com/ivan-lednev/obsidian-day-planner/files/13695417/demo-vault-2.tar.gz)

The "switching week" performance in your version went down linearly (from ~210 to ~430 ms), which makes sense; `useVisibleTasks` has to plow through all the data when the visible days change. I doubled the data once more (16,000 notes, 40,000 items per type) to verify this and yours was at ~950 ms then. My optimized version stayed within the same numbers regardless of the size of the vault.

You probably can apply these optimizations to your code as well, but it still wouldn't support list items then. Getting this back was, after all, my main goal, because you wrote:

> The plugin now pulls tasks from Dataview cache, and including list items would make this very slow on large vaults or in vaults where everything is a list item (I personally write everything as list items)

These numbers don't look too bad to me, but let me know if you think I've missed something.

weph commented:
Any news on this, @ivan-lednev ?

ivan-lednev commented:
Hello, @weph,

I finally have a bit more time to work on the project, so I've come up with the following: if the Dataview filter is empty, only pull tasks from _visible_ daily notes (instead of 'all over the vault'). Since most users use only daily notes anyway, this completely fixes the issue for them. The results are now sub-millisecond. 

I drew inspiration from your solution, so currently it looks like this:
https://github.com/ivan-lednev/obsidian-day-planner/commit/6d6b596ff6c7f36cf991cfb68c6baa5d42a32a59#diff-e1b03dbd975c40e85c93ae0c2fade1ac3bc6bc211df837b7b81ca1f3e5744fc3R26

I'm testing it in my own workflow, so once I have the time to fix some annoying issues, it's good to go.

Let me know what you think!

weph commented:
Hey @ivan-lednev 

thanks for the update. I just installed the latest version and it looks great! I haven't looked into your implementation yet, but I'm glad to hear my work contributed to finding a solution you're happy with.
